File: src/go/problems/min_distance_test.go
// https://leetcode.com/problems/edit-distance

package problems

import (
	"sort"
	"testing"
	"math"
	"github.com/dobermanch/leetcode/core"
)

type MinDistance struct{}

func TestMinDistance(t *testing.T) {
	gen := core.TestSuite[MinDistance]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("horse").Param("ros").Result(3)
	}).Add(func(tc *core.TestCase) {
		tc.Param("intention").Param("execution").Result(5)
	}).Run(t)
}

func (MinDistance) Solution(word1 string, word2 string) int {
    minFunc := func(left int, right int) int {
        if left < right {
            return left
        }
        return right
    }

    height := len(word1)
    width := len(word2)

    result := make([][]int, height + 1)

    for i := 0; i <= height; i++ {
        result[i] = make([]int, width + 1)
        result[i][0] = i
    }

    for i := 0; i <= width; i++ {
        result[0][i] = i;
    }

    for i := 1; i <= height; i++ {
        for j := 1; j <= width; j++ {
            add := 0
            if word1[i - 1] != word2[j - 1] {
                add = 1
            }
            
            min := minFunc(result[i - 1][j] + 1, result[i][j - 1] + 1)
            result[i][j] = minFunc(min, result[i - 1][j - 1] + add)
        }
    }

    return result[height][width]
}

File: src/python/problems/min_distance_test.py
#https://leetcode.com/problems/edit-distance

from core.problem_base import *

class MinDistance(ProblemBase):
    def Solution(self, word1: str, word2: str) -> int:
        map = [[0] * (len(word2) + 1) for _ in range(len(word1) + 1)]

        for i in range(len(word1) + 1):
            map[i][0] = i
        
        for i in range(len(word2) + 1):
            map[0][i] = i

        for i in range(1, len(word1) + 1):
            for j in range(1, len(word2) + 1):
                add = 0 if word1[i - 1] == word2[j - 1] else 1
                map[i][j] = min(map[i - 1][j] + 1, map[i][j - 1] + 1, map[i - 1][j - 1] + add)

        return map[-1][-1]

if __name__ == '__main__':
    TestGen(MinDistance) \
        .Add(lambda tc: tc.Param("word1", "horse").Param("word2", "ros").Result(3)) \
        .Add(lambda tc: tc.Param("word1", "intention").Param("word2", "execution").Result(5)) \
        .Run()
